### ISSUE TITLE
Compostionality of Atomese processing stages (was "StopUsing SetLinks!)

### ISSUE THREAD

**Issue by linas** [opencog/atomspace#2911]
Issue #1502 "***Stop using SetLink for search results!***" highlights a problem with the compositionality of the search and processing primitives in Atomese. Search results were (still are) delivered wrapped in a `SetLink`.  This presents challenges for the cleanup of search results, and various other core issues (See [issues described on the SetLink wiki](https://wiki.opencog.org/w/SetLink).)

Basically, it is difficult to create long processing pipelines, such as `(cog-evaluate! (SequentialAnd ... (Put .... (State ... (Get ... (SequentialOr ... (Delete ... (Bind ...` with the top-level `SequentialAnd` written to be tail recursive. Such long pipelines have been implemented (in the robot code, circa 2017) and they do work (see for example the [tail-recusion demo](https://github.com/opencog/atomspace/blob/master/examples/atomspace/recursive-loop.scm) and older copies of the https://github.com/opencog/ros-behavior-scripting repo)  However, there are problems:

* There is difficulty in deleting the wrapper `SetLink`s when they are no longer needed. 
* Writing long chained pipelines seems more difficult than it should be.
* Awkward internal implementation of how SetLinks are handled when they are passed to `PutLink` and other assorted functions (e.g. `PlusLink` when given a set of numbers  to add.)

A replacement solution is needed.  Desirable properties:
* Ability to get search results incrementally, as they come in, rather than getting one big blob at the end.
* Ability to run one query in parallel mode.
* Eliminate the need for, or replace [`MapLink`](https://github.com/opencog/atomspace/blob/master/examples/pattern-matcher/map.scm)

A wholly unexplored idea:
* **Monads** -- The PutLink and other misc links currently accept either individual atoms, or sets of atoms, as input. In the case of sets, the contained atoms are automatically unwrapped and processed.  This suggests that the current use of `Setlink` is as a kind of sloppy [monad](https://en.wikipedia.org/wiki/Monad_(functional_programming)) and thus what we really need is a cleaned-up monad to hold multiple-value results. The precise way to do this is unclear

Another idea, inspired by the ROS work:
* **Generic Publish-Subscribe system** See #1750 --  Maybe compositionaliity should be achieved by the ability to publish and subscribe to results? Some kind of streams, promises, futures subsystem?

The following attempts have been made:
* **Anchor proposal** -- This suggests specifying an AnchorLink in the query, and chaining on results with MemberLink as they show up. This has been implemented and documented and unit-tested. See #2500 No one uses it. There has not been any exploration at all of how  compositionality works.
* **QueryLink returning QueueValue** -- implemented, works, documented, unit tested. See  #2571  How do do compositionality with this is unexplored.

Exploring compositionality with these two styles seems to require:
* Semaphores or locks (i.e. parts of the publish-subscribe subsystem). 
* Some form of lazy evaluation or of [`$vau` (aka fexpers)]( https://en.wikipedia.org/wiki/Fexpr) beyond the simplistic form offered by `SequentialAndLink`

Possible building block for monads or related:
* **AtomSpaceNode** -- This suggests treating the AtomSpace as a kind of mutable Link. Thus, the query would return an AtomSpace, layered on the main space, holding the search results. Partly implemented: See #2865 The `AtomSpacePtr` is now a kind of `ValuePtr`. No one is using this for anything, just yet. There is no way to automatically project/collapse contents of derived atomspaces back into the main atomspace. (You'd have to do it by hand.)  This could generalize: Atomspaces could be layered arbitrarily deep; passed around, hold temporary results that disappear when all references to the AtomSpace disappear.  AtomSpaces could be stored in ordinary Links.
* ** QueueValue** -- This is currently used to hold search results from `QueryLink` and `MeetLink`. The whole value flow subsystem is envisioned as being able to handle flows of  ... values ... and not flows of atoms.

To explore these issues, and possible solutions, some demos are suggested.  So far, we have the following demos:
* **dot product** -- Demo of how to compute the dot-product of two vectors, where the two vectors are sets of atoms returned by a searrch query. See [dot-product.scm](https://github.com/opencog/atomspace/blob/master/examples/pattern-matcher/dot-product.scm)
* **recursive query** -- Demo of how a recursive query can be written. See [recursive.scm](https://github.com/opencog/atomspace/blob/master/examples/pattern-matcher/recursive.scm)

**Comment by linas** [comment]
At this time, building on top of `QueueValue` seems the most sensible thing to do. This could be formalized by providing an API within QueuueValue itself, instead of inheriting from what's in cogutils.

**Comment by linas** [comment]
The following attempt was made:
* **Anchor proposal** -- This suggests that an AnchorLink can be specified in the query, and results are chained on with MemberLink, as they show up. This has been implemented and documented and unit-tested. See #2500 No one uses it. It was reverted earlier today, in eee7a61ea766d6f94d9b1ef3a2caab649d793ba4

There has been some minor exploration of how  compositionality works; it is demoed in the (still existing) example [query.scm](https://github.com/opencog/atomspace/tree/master/examples/pattern-matcher/query.scm)  This example uses AnchorNodes, but without needing pull req #2500 to do it. It "works". It's even multi-threaded, so its "naturally" parallel. Is it clunky? I dunno. It posts results to the AtomSpace, ... but why? was this really needed?
